src/messages.rs

```diff
@@ -194,49 +202,97 @@ impl HopMessage {
     pub fn content(&self) -> &SignedMessage {
         &self.content
     }
+}
 
-    pub fn route(&self) -> u8 {
-        self.route
-    }
-
-    pub fn sent_to(&self) -> &Vec<XorName> {
-        &self.sent_to
-    }
+/// A list of a group's public IDs, together with a list of signatures of a neighbouring group.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Clone, Hash, RustcEncodable, RustcDecodable)]
+pub struct GroupList {
+    // TODO(MAID-1677): pub signatures: BTreeSet<(PublicId, sign::Signature)>,
+    pub pub_ids: BTreeSet<PublicId>,
 }
 
 /// Wrapper around a routing message, signed by the originator of the message.
 #[derive(Ord, PartialOrd, Eq, PartialEq, Clone, Hash, RustcEncodable, RustcDecodable)]
 pub struct SignedMessage {
     /// A request or response type message.
     content: RoutingMessage,
-    /// Claimed public ID of a node or client.
-    ///
-    /// For clients this is easily verifiable since their name is computed from the ID. For nodes it
-    /// needs to be confirmed by their `NodeManager`.
-    public_id: PublicId,
-    signature: sign::Signature,
+    /// The lists of the groups involved in routing this message, in chronological order.
+    grp_lists: Vec<GroupList>,
+    /// The IDs and signatures of the source authority's members.
+    signatures: BTreeMap<PublicId, sign::Signature>,
 }
 
 impl SignedMessage {
     /// Creates a `SignedMessage` with the given `content` and signed by the given `full_id`.
     pub fn new(content: RoutingMessage, full_id: &FullId) -> Result<SignedMessage, RoutingError> {
-        let bytes_to_sign = try!(serialise(&(&content, full_id.public_id())));
+        let sig = sign::sign_detached(&try!(serialise(&content)), full_id.signing_private_key());
         Ok(SignedMessage {
             content: content,
-            public_id: *full_id.public_id(),
-            signature: sign::sign_detached(&bytes_to_sign, full_id.signing_private_key()),
+            grp_lists: Vec::new(),
+            signatures: iter::once((*full_id.public_id(), sig)).collect(),
         })
     }
 
-    /// Confirms the signature against the claimed public ID.
+    /// Confirms the signatures.
     pub fn check_integrity(&self) -> Result<(), RoutingError> {
-        let signed_bytes = try!(serialise(&(&self.content, &self.public_id)));
-        if sign::verify_detached(&self.signature,
-                                 &signed_bytes,
-                                 self.public_id().signing_public_key()) {
-            Ok(())
-        } else {
-            Err(RoutingError::FailedSignature)
+        let signed_bytes = try!(serialise(&self.content));
+        for (pub_id, sig) in &self.signatures {
+            if !sign::verify_detached(sig, &signed_bytes, pub_id.signing_public_key()) {
+                return Err(RoutingError::FailedSignature);
```

Comment:
Yes, you're right, a bad signature should not stop us from handling the message.